src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp

```diff
@@ -0,0 +1,180 @@
+    }
+
+    auto metadata_snapshot = storage.getInMemoryMetadataPtr();
+
+    const auto & replica_path = storage.replica_path;
+    /// May it be ZK lost not the whole root, so the upper check passed, but only the /replicas/replica
+    /// folder.
+    bool replica_path_exists = zookeeper->exists(replica_path);
+    if (!replica_path_exists)
+    {
+        LOG_WARNING(log, "No metadata in ZooKeeper for {}: table will stay in readonly mode", replica_path);
+        storage.has_metadata_in_zookeeper = false;
+        return;
+    }
+
+    storage.has_metadata_in_zookeeper = true;
+
+    /// In old tables this node may missing or be empty
+    String replica_metadata;
+    const bool replica_metadata_exists = zookeeper->tryGet(replica_path + "/metadata", replica_metadata);
+
+    if (!replica_metadata_exists || replica_metadata.empty())
+    {
+        /// We have to check shared node granularity before we create ours.
+        storage.other_replicas_fixed_granularity = storage.checkFixedGranularityInZookeeper();
+
+        ReplicatedMergeTreeTableMetadata current_metadata(storage, metadata_snapshot);
+
+        zookeeper->createOrUpdate(replica_path + "/metadata", current_metadata.toString(), zkutil::CreateMode::Persistent);
+    }
+
+    storage.checkTableStructure(replica_path, metadata_snapshot);
+    storage.checkParts(skip_sanity_checks);
+
+    if (zookeeper->exists(replica_path + "/metadata_version"))
+    {
+        storage.metadata_version = parse<int>(zookeeper->get(replica_path + "/metadata_version"));
+    }
+    else
+    {
+        /// This replica was created with old clickhouse version, so we have
+        /// to take version of global node. If somebody will alter our
+        /// table, then we will fill /metadata_version node in zookeeper.
+        /// Otherwise on the next restart we can again use version from
+        /// shared metadata node because it was not changed.
+        Coordination::Stat metadata_stat;
+        zookeeper->get(zookeeper_path + "/metadata", &metadata_stat);
+        storage.metadata_version = metadata_stat.version;
+    }
+
+    /// Temporary directories contain uninitialized results of Merges or Fetches (after forced restart),
+    /// don't allow to reinitialize them, delete each of them immediately.
+    storage.clearOldTemporaryDirectories(0, {"tmp_", "delete_tmp_", "tmp-fetch_"});
+    storage.clearOldWriteAheadLogs();
+    if (storage.getSettings()->merge_tree_enable_clear_old_broken_detached)
+        storage.clearOldBrokenPartsFromDetachedDirecory();
+
+    storage.createNewZooKeeperNodes();
+    storage.syncPinnedPartUUIDs();
+
+    storage.createTableSharedID();
+};
+
+void ReplicatedMergeTreeAttachThread::finalizeInitialization() TSA_NO_THREAD_SAFETY_ANALYSIS
+{
+    std::unique_lock lock(storage.initialization_mutex);
+    if (storage.startup_called)
+    {
+        lock.unlock();
+        storage.startupImpl();
+        lock.lock();
+    }
+
+    storage.initialization_done = true;
```

Comment:
Do I understand correctly, that the purpose of this code is to call `startupImpl` exactly once: either from the thread that loads tables or from AttachThread? Maybe we can always call `startupImpl` from AttachThread when finalizing initialization?

Reply:
yes, that's correct
And I 100% agree, with your comment below, that will make it even simpler.
I was too focused on keeping it the same (attaching in constructor) and just assumed that doing it during startup isn't an option. If you don't see any problem with it, I'll start the thread and wait for the first try in startup.